File: src/main/java/org/gautelis/durga/plugins/FieldFilter.java
package org.gautelis.durga.plugins;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.Arrays;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Whitespace or blacklist filtering on JSON payload fields.
 *
 * <p>If {@code keep} is specified, only those fields (including nested via dot-notation)
 * are retained. If {@code drop} is specified, those fields are removed. If both are
 * specified, {@code keep} wins on conflict.
 */
public final class FieldFilter implements Plugin {

    @Override
    public String execute(String payload, String config) throws Exception {
        String keep = null;
        String drop = null;
        String flatten = null;
        if (config != null && !config.isBlank()) {
            String[] parts = config.split("\\s+");
            for (String part : parts) {
                int eq = part.indexOf('=');
                if (eq > 0) {
                    String key = part.substring(0, eq).trim();
                    String val = part.substring(eq + 1).trim();
                    switch (key) {
                        case "keep" -> keep = val;
                        case "drop" -> drop = val;
                        case "flatten" -> flatten = val;
                    }
                }
            }
        }
        return filter(payload, keep, drop, flatten);
    }

    public FieldFilter() {
    }

    /**
     * Filters fields from a JSON payload.
     *
     * @param json input JSON string
     * @param keep comma-separated fields to retain (optional)
     * @param drop comma-separated fields to remove (optional)
     * @param flattenPrefix if non-null, hoist nested fields under this prefix to top level
     * @return filtered JSON string
     */
    public static String filter(String json, String keep, String drop, String flattenPrefix) {
        ObjectMapper mapper = new ObjectMapper();
        JsonNode input;
        try {
            input = mapper.readTree(json);
        } catch (JsonProcessingException e) {
            throw new IllegalArgumentException("Invalid input JSON", e);
        }
        if (!input.isObject()) {
            return json;
        }

        Set<String> keepSet = keep != null && !keep.isBlank()
                ? Arrays.stream(keep.split(",")).map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toSet())
                : null;
        Set<String> dropSet = drop != null && !drop.isBlank()
                ? Arrays.stream(drop.split(",")).map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toSet())
                : null;

        ObjectNode output = mapper.createObjectNode();
        Iterator<Map.Entry<String, JsonNode>> fields = input.properties().iterator();
        while (fields.hasNext()) {
            Map.Entry<String, JsonNode> entry = fields.next();
            String field = entry.getKey();

            boolean shouldKeep;
            if (keepSet != null && !keepSet.isEmpty()) {
                shouldKeep = keepSet.contains(field);
            } else if (dropSet != null && dropSet.contains(field)) {
                shouldKeep = false;
            } else {
                shouldKeep = true;
            }
            if (shouldKeep) {
                output.set(field, entry.getValue());
            }
        }

        if (flattenPrefix != null && !flattenPrefix.isBlank()) {
            JsonNode nested = input.get(flattenPrefix);
            if (nested != null && nested.isObject()) {
                Iterator<Map.Entry<String, JsonNode>> nestedFields = nested.properties().iterator();
                while (nestedFields.hasNext()) {
                    Map.Entry<String, JsonNode> entry = nestedFields.next();
                    output.set(entry.getKey(), entry.getValue());
                }
            }
        }

        return output.toString();
    }
}

File: durga-rust/src/plugins/field_filter.rs
//! Port of `org.gautelis.durga.plugins.FieldFilter` (plugin id `field-filter`).
//!
//! Retains (`keep`) or removes (`drop`) top-level fields, and optionally hoists
//! (`flatten`) a nested object's fields to the top level. `keep` wins over
//! `drop` on conflict.

use std::collections::HashSet;

use serde_json::{Map, Value};

use crate::plugin::{Plugin, PluginError};

#[derive(Default)]
pub struct FieldFilter;

impl FieldFilter {
    pub fn new() -> Self {
        FieldFilter
    }

    pub fn filter(json: &str, keep: Option<&str>, drop: Option<&str>, flatten_prefix: Option<&str>)
        -> Result<String, PluginError>
    {
        let input: Value = serde_json::from_str(json)
            .map_err(|e| -> PluginError { format!("Invalid input JSON: {e}").into() })?;
        let Value::Object(obj) = &input else {
            return Ok(json.to_string());
        };

        let keep_set = split_set(keep);
        let drop_set = split_set(drop);

        let mut output = Map::new();
        for (field, value) in obj {
            let should_keep = if let Some(keep) = &keep_set {
                !keep.is_empty() && keep.contains(field.as_str())
            } else {
                drop_set.as_ref().map(|d| !d.contains(field.as_str())).unwrap_or(true)
            };
            if should_keep {
                output.insert(field.clone(), value.clone());
            }
        }

        if let Some(prefix) = flatten_prefix.filter(|p| !p.is_empty()) {
            if let Some(Value::Object(nested)) = obj.get(prefix) {
                for (k, v) in nested {
                    output.insert(k.clone(), v.clone());
                }
            }
        }

        Ok(Value::Object(output).to_string())
    }
}

fn split_set(spec: Option<&str>) -> Option<HashSet<String>> {
    spec.filter(|s| !s.trim().is_empty()).map(|s| {
        s.split(',')
            .map(|f| f.trim().to_string())
            .filter(|f| !f.is_empty())
            .collect()
    })
}

impl Plugin for FieldFilter {
    fn execute_str(&self, payload: &str, config: &str) -> Result<Option<String>, PluginError> {
        let mut keep = None;
        let mut drop = None;
        let mut flatten = None;
        for part in config.split_whitespace() {
            let Some(eq) = part.find('=') else { continue };
            if eq == 0 {
                continue;
            }
            let key = part[..eq].trim();
            let val = part[eq + 1..].trim();
            match key {
                "keep" => keep = Some(val.to_string()),
                "drop" => drop = Some(val.to_string()),
                "flatten" => flatten = Some(val.to_string()),
                _ => {}
            }
        }
        Ok(Some(Self::filter(payload, keep.as_deref(), drop.as_deref(), flatten.as_deref())?))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn keep_retains_only_listed() {
        let out = FieldFilter::new()
            .execute_str(r#"{"a":1,"b":2,"c":3}"#, "keep=a,c")
            .unwrap()
            .unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert!(v.get("a").is_some() && v.get("c").is_some() && v.get("b").is_none());
    }

    #[test]
    fn drop_removes_listed() {
        let out = FieldFilter::new()
            .execute_str(r#"{"a":1,"b":2}"#, "drop=b")
            .unwrap()
            .unwrap();
        let v: Value = serde_json::from_str(&out).unwrap();
        assert!(v.get("a").is_some() && v.get("b").is_none());
    }
}
